mantidproject/mantid
# Run Systemtests Using a `ctest` Wrapper Around Our Custom System Testing Implementation

### Description of work

#### Summary of work
- Creates a much more streamlined wrapper around our `systemtesting` imeplementation that allows it to be primarily powered by `ctest`. 
- Rather than using the `TestManager` class in `systemtesting.py` that handles the setup of each test, ctest will handle things like test selection and then hand off to `systestrunner` which creates the `TestScript` object and a `TestRunner` to run it.
- Running only system tests or only unit tests can be performed by passing a label to `ctest`: `-L UnitTest` or `-L SystemTest`
  - Not including a label allows for finding and running of sets of both Unit and System test. E.g: you modify a SANS algorithm and can now run `ctest -R SANS` to make sure all SANS tests are still passing before sending up to the CI. 
- The motivation behind this is improving the similarity between the running of our UnitTest and SystemTest suites. This should simplify the way our CI is forced to interact with these tests - only requiring calls to `ctest`. It also removes the need for both the `systemtest` and `systemtest.bat` shell and batch scripts.

Fixes #34850.

### To test:
- Run a subset of unit tests using: `ctest -R <whatever> -L UnitTest`
- Run a subset of system tests using `ctest -R <whatever> -L SystemTest`
- Run a subset of all tests using `ctest -R <whatever>`
- Ensure that `systemtest -R <whatever>` still works. 


*This does not require release notes* because **it concerns testing only.**

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
